sdk/core/Azure.Core/src/Serialization/AzureObjectSerializer.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// An abstraction from reading typed objects.
+    /// </summary>
+    public abstract class AzureObjectSerializer
+    {
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract void Serialize(Stream stream, object? value, Type inputType);
+
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract ValueTask SerializeAsync(Stream stream, object? value, Type inputType);
+
+        /// <summary>
+        /// Read the binary representation into a <paramref name="returnType"/>.
+        /// The Stream will be read to completion.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to read from.</param>
+        /// <param name="returnType">The type of the object to convert to and return.</param>
+        public abstract object Deserialize(Stream stream, Type returnType);
```

Comment:
One scenario Cosmos DB .NET currently has is an issue with abstractions. Where users ask to serialize a list of object based on abstract base class which works, but then the deserialize fails because it can't create an instance of the abstract class.

```
public abstract class Animal { } 
List<Animal> animals = new List<Animal>() { new Cat(), new Dog() };
MemoryStream stream = new MemoryStream();
await serializer.SerializeAsync(stream, animals, typeof(List<Animal>));
List<Animal> output = await serializer.DeserializeAsync(stream, typeof(List<Animal>));
```

Reply:
Are you using the System.Text.Json serializer?

Reply:
Yeah, we have reports of it happening with System.Text.Json https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1255.
I tested it with System.Text.Json 4.7.1 and it throws `Deserialization of reference types without parameterless constructor is not supported`.

```
class Program
    {
        static void Main(string[] args)
        {
            Dog dog = new Dog("john");
            string serialized = JsonSerializer.Serialize(dog);
            Animal animal1 = JsonSerializer.Deserialize<Animal>(serialized);
        }
    }


    public abstract class Animal
    {
        public Animal()
        {
        }

        public string Name { get; set; }

        public abstract void Do();
    }

    public class Dog : Animal
    {
        public Dog(string name)
        {
            this.Name = name;
        }

        public override void Do()
        {
            Console.WriteLine("Bark");
        }
    }
```

Reply:
It's a good issue, but is it relevant to the abstraction design? I think implementation of the abstraction simply need to know how to map from abstract types passed to the deserialize method to concrete types that the deserializer should instantiate. Different concrete serializers do this differently.